src/Model/Post.php

```diff
@@ -0,0 +1,218 @@
+<?php
+/**
+ * @copyright Copyright (C) 2020, Friendica
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace Friendica\Model;
+
+use Friendica\Core\Protocol;
+use Friendica\Database\DBA;
+use Friendica\Protocol\Activity;
+use Friendica\Repository\PermissionSet;
+
+class Post
```

Comment:
Great, a new class, this is the perfect opportunity to apply the Domain Driven Design principles and have more testable code, especially for the critical post features!

This means that we would need two non-static classes:
- A `Post` repository class which constructor takes a Database instance which it will use to perform database queries
- A `Post` model which constructor will take all the fields you've described in the view

The Post repository class will perform the SELECT on the views and then instantiate Post model class from the data.

The Post model class will perform internal data changes but no database calls will be issued from it.

The Post repository will then persist the changes through an UPDATE method taking one (or several) Post models as parameter.

Let me know if you need more information or guidance through this process.

Reply:
I don't understand a word at all, I must confess.

And just to clarify: I really would dislike when in the future that class would had to be called non statically or by performing some weird workarounds.

Reply:
I know you wouldn't like it, but it isn't just about you. Right now you are showing an exceptional commitment to Friendica in terms of time spent, both implementing new protocols, troubleshooting weird behaviors and improving performances. And it's great! But if you stop working on the project for some reason (personal, medical, romantic), we will be left with a codebase that is mostly revolving around your own development paradigm: static classes and barely any tests.

Of course, if there was some formal way of ensuring you would be continuing to work on Friendica forever, then we would be able to blindly follow your preferences, but unfortunately it isn't the case. Unit tests are a way to ensure behavior consistency even in your absence, and non-static decoupled classes are the way to go. Of course I'll be assisting you during what I feel is a necessary transition every way I can.

And, to be completely honest, unit tests are pretty new and uncomfortable for me too! But I believe they are important for the project health as a whole. It makes it more resilient to code changes and reduce the amount of time we spend chasing weird behaviors. I like the idea of automatically catching mistakes we keep making before a user reports an issue. Of course it doesn't always show in the PRs, but the existing tests have already helped me correcting mistakes before it was merged.

Reply:
Thing is: The database part of the `Item` class is called at around 400 places (or more) in the whole code. If the replacement wouldn't be 100% parameter compatible (including the return values), replacing it would be a task of several months of work - without any possibility of doing any bugfixing or any other - even small - change in the codebase. We haven't even achieved replacing each of the really old `q(...)` calls by now - and this would be a comparable small task.

I completely do understand the motivation behind the changes. But the transition shouldn't be done with the biggest and most used class in the system - and it should be done in a way that I do understand as well. For example the notification class is done in a newer way - and I still struggle with even understanding what's going on in that class or how to use it - and I wouldn't dare performing any serious changes in that class at all.

Reply:
I hear you, but the notification class architecture, from @nupplaphil 's own admission, isn't the best example of DDD. Please don't let it deter you from the philosophy it doesn't do a good job at presenting.

Nevertheless, you are right about the migration priority, we probably should replace all the `q()` calls before we can get fancier. I was hoping we could skip a step but my impatience is probably making me want us to run before we walk.

Reply:
Concerning all database classes I would prefer a solution where the class variables that represent the table fields are created automatically. We already have got all data in the table structure specification, so we should automatically generate and update the code based upon this. Doing so would bring us to a situation where these classes would be available for all tables instantly.

BTW: Some unsolved problem with using a class that contains all fields would be the performance. Especially with this view here the difference is significant. Fetching all fields of the view instead of only the common ones according to my tests is more than 3 times slower. Means that - for example - the network page would load in 6 seconds instead of 2, also post delivery would slow done by this factor of 3. MySQL seems to optimize the query for the view depending upon which fields are selected. This is a problem that we have to touch then as well.

Reply:
Ah, interesting.

Reply:
I guess (hope) that this speed difference is specific to views and hopefully shouldn't affect regular queries on a single table.